### PR TITLE
Explore uncorrelated generic genes

### PR DESCRIPTION
**Pull request by ajlee21**
This PR performs an analysis to examine the genes in the bottom right of the [correlation plot](https://github.com/greenelab/generic-expression-patterns/blob/master/human_general_analysis/gene_ranking_log2FoldChange.svg)

In this case, SOPHIE was trained on recount2 (RNA-seq) dataset while the manually curated dataset was using array platform.

The goal of these analyses is to examine why these genes are uncorrelated (i.e. generic in RNA-seq and not array). Specifically, we believe that these RNA-seq generic genes are due to differences in platform so we will test two different hypotheses that get are this.

**Hypothesis 1:** These RNA-seq generic genes are due to difference in platforms. Specifically, we believe that these genes are those that are differentially expressed using ribo-depleted vs polyA-selection. There may exist some genes that are frequently found to be differentially expressed using polyA-selection and some genes that are frequently found to be differentially expressed using the ribo-depletion strategy. Since our training compendium is recount2, which includes many different types of experiments therefore different selection strategies, its possible that these gene biases that result from differences in processing are found by SOPHIE but not found by Crow et. al. which uses array technology.

**Results:** 
image

DEGs are highlighted in red on top of the correlation plot are not localized to the bottom right, so these RNA-seq generic genes don't seem to be the results of differences in protocol.

**Hypothesis 2:** These RNA-seq generic genes are those that are not well captured on microarray technology. [Zhao et. al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3894192/) found that one possible reason for differences in differentially expressed genes detected on different platforms is due to how genes are captured. RNA-Seq is more likely to detect the changes at two different conditions for genes with very low expression or very high expression compared to arrays.

**Results:**  
RNA-seq expression
image

Array expression
image

On the array, it looks like RNA-seq generic genes are fairly lowly expressed, but based on the density of the violin plot there appear to be many genes that have a similar range of expression, include RNA-seq and array generic genes. .So these RNA-seq generic genes are as well captured as RNA-seq and array generic genes. This doesn't support our hypothesis 2.

However, we noticed that using RNA-seq, it looks like the RNA-seq generic genes are very lowly expressed and distinct compared to the RNA-seq and array generic genes. This might point us to something else we can look into.

### CODE REVIEW

**Review by jjc2718 — APPROVED** [review]
Nice work! Code changes look good, most of my comments are small stylistic things.

For the array expression violin plots, it does look to me like the "only RNAseq generic" genes overall have lower expression - were you expecting to see a clearer separation between that and the "RNA-seq and array generic" distribution? They definitely don't seem randomly sampled to me, but I'm assuming you and Casey have talked this through and you were looking for something else.

It is interesting that they're lowly expressed in the RNA-seq data too - do you have any ideas for what the next steps would be for that?

File: explore_RNAseq_only_generic_genes/0a_process_PBTA_data.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+# +
+# Select patient sample ids with both polyA-selected and ribo-depleted measurements
+shared_patient_ids = list(
+    set(patient_to_polya_id.keys()).intersection(patient_to_ribo_id.keys())
+)
+
+# Check that these patient ids were consistent with previous analysis comparing TP53 status across platform:
+# https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/930
+shared_patient_ids
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply explore_RNAseq_only_generic_genes/0a_process_PBTA_data.py:135]: Yes this was a manual check. I will update this comment

File: explore_RNAseq_only_generic_genes/1a_DE_polyA_vs_ribo.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+)
+
+params = utils.read_config(config_filename)
+local_dir = params["local_dir"]
+dataset_name = params["dataset_name"]
+col_to_rank_genes = params["rank_genes_by"]
+project_id = params["project_id"]
+# -
+
+# ## Differential expression analysis
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply explore_RNAseq_only_generic_genes/1a_DE_polyA_vs_ribo.py:54]: Yea this took a little bit of digging but I'm glad I did it. I definitely understand better what the input data is!

File: explore_RNAseq_only_generic_genes/README.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Explore RNA-seq generic only genes
+
+When we compared the correlation between gene percentiles generated by SOPHIE versus the manually curated dataset [here](https://github.com/greenelab/generic-expression-patterns/blob/master/human_general_analysis/gene_ranking_log2FoldChange.svg), we noticed that there was a group of genes that SOPHIE identified as generic but were not found to be generic using the manually curated dataset. In this case, SOPHIE was trained on recount2 (RNA-seq) dataset while the manually curated dataset was using array platform.
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply explore_RNAseq_only_generic_genes/README.md:3]: Good point. Something that we do for manubot, which I'm sure is also where you learned this too, but I have forgotten to do here

**Comment by ajlee21** [comment]
> For the array expression violin plots, it does look to me like the "only RNAseq generic" genes overall have lower expression - were you expecting to see a clearer separation between that and the "RNA-seq and array generic" distribution? They definitely don't seem randomly sampled to me, but I'm assuming you and Casey have talked this through and you were looking for something else.

Our hypothesis was that these "only RNAseq generic" genes might be found to be DE in the RNA-seq platform because they are genes that are poorly captured on the array platform (i.e. we'd expect to see lower activity of these genes on the array). However, we found that, in the array dataset, "only RNAseq generic" genes have similar expression as RNA-seq/array generic genes. So it doesn’t look like the RNA-seq generic genes are found due to poor array capture

> It is interesting that they're lowly expressed in the RNA-seq data too - do you have any ideas for what the next steps would be for that?

Through this analysis, we noticed that RNA-seq generic genes are distinctly low compared to RNA-seq/array generic genes in RNA-seq data. There are a couple of possibilities here:

* These RNA-seq generic genes might be considered noise in array experiments, but are detected in DE using RNA-seq, which is more sensitive
* Could also be that VAE isn’t adequately accounting for low counts. So say we have an experiment with low read depth where a set of genes have 0 expression (i.e. not detectable). After going through the VAE shifting process, the gene counts might get compressed (i.e. very low gene counts are increased and very high gene counts are decreased) so that the new simulated counts for these originally 0-expressed genes, which should have a high error rate and therefore not found to be DE by DESeq, can be found as DE by DESeq with the artificial shift in values. If this is the case, we will need a way to re-scale values to account for the read depth differences. This will be something to test in the next PR

File: generic_expression_patterns_modules/ranking.py [hunk 4]
```diff
@@ -549,6 +546,87 @@ def compare_gene_ranking(
...
+        kind='hex',
+        marginal_kws={"color": "white"},
+    )
+    highlight_data = shared_gene_rank_scaled_df[shared_gene_rank_scaled_df["gene label"] == "polyA_vs_ribo"]
+
+    # print(highlight_data)
+
+    plt.scatter(highlight_data["Percentile (simulated)"], highlight_data[ref_rank_col], c="red", alpha=0.7)
+
+    # fig = sns.jointplot(
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply generic_expression_patterns_modules/ranking.py:599]: yup I can remove these, leftover from testing

File: explore_RNAseq_only_generic_genes/1a_DE_polyA_vs_ribo.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+#
+# In our case, since RSEM is returning gene-level summaries, `tximport` will just combines the gene-level summaries across files into a matrix.
+#
+# The dataset we have is a matrix of estimated counts that were combined across samples. We don't have the individual raw data files output from RSEM, which is needed to call tximport. However, it looks like `tximport` would have been redundant given what we mentioned above. Here we are rounding the combined RSEM estimated counts and using `DESeqDataSetFromMatrix()`. Given the first step of `DESeqDataSetFromTximport` is rounding the counts, we anticipate this step to be ok. Looks like rounding is also performed for `DESeqDataSet()` as well: https://rdrr.io/bioc/DESeq2/src/R/AllClasses.R.
+#
+# In general, the recommendation to use `tximport` might be necessary for the case where we have transcript-level data.
+
+# +
+mapped_expression_filename = "polya_ribo_expression.tsv"
+processed_expression_filename = "polya_ribo_expression_processed.tsv"
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply explore_RNAseq_only_generic_genes/1a_DE_polyA_vs_ribo.py:121]: No, this processing is to prepare the data for input into DESeq. This processing includes:
* checking and reordering samples ids according to metadata grouping
* rounding estimated counts to integers
I'll add a comment here to make that clear

**Comment by ajlee21** [comment]
Pytests are still failing for the same reason here: https://github.com/greenelab/generic-expression-patterns/issues/69

I have commented out the pytest calls in my `tests.yaml` file temporarily and will resolve this issue later. I feel ok about doing this since the tests run fine locally, so this seems to be an issue with github.

**Pull request merged**